telethon/full_sync.py

```diff
@@ -0,0 +1,143 @@
+    try:
+        res = await af
+        cf.set_result(res)
+    except Exception as e:
+        cf.set_exception(e)
+
+
+def _sync_result(loop, x):
+    f = Future()
+    loop.call_soon_threadsafe(asyncio.ensure_future, _proxy_future(x, f))
+    return f.result()
+
+
+class _SyncGen:
+    def __init__(self, loop, gen):
+        self.loop = loop
+        self.gen = gen
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        try:
+            return _sync_result(self.loop, self.gen.__anext__())
+        except StopAsyncIteration:
+            raise StopIteration from None
+
+
+def _syncify_wrap(t, method_name, loop, thread_ident, syncifier=_sync_result):
+    method = getattr(t, method_name)
+
+    @functools.wraps(method)
+    def syncified(*args, **kwargs):
+        coro = method(*args, **kwargs)
+        return (
+            coro if threading.get_ident() == thread_ident
+            else syncifier(loop, coro)
+        )
+
+    setattr(t, method_name, syncified)
+
+
+def _syncify(*types, loop, thread_ident):
+    for t in types:
+        for method_name in dir(t):
+            if not method_name.startswith('_') or method_name == '__call__':
+                if inspect.iscoroutinefunction(getattr(t, method_name)):
+                    _syncify_wrap(t, method_name, loop, thread_ident, _sync_result)
+                elif isasyncgenfunction(getattr(t, method_name)):
+                    _syncify_wrap(t, method_name, loop, thread_ident, _SyncGen)
+
+
+__asyncthread = None
+
+
+def enable(loop=None, executor=None, max_workers=1):
+    global __asyncthread
+    if __asyncthread is not None:
+        raise RuntimeError("full_sync can only be enabled once")
+
+    if not loop:
+        loop = asyncio.get_event_loop()
+    if not executor:
+        executor = ThreadPoolExecutor(max_workers=max_workers)
+
+    def start():
+        asyncio.set_event_loop(loop)
+        loop.run_forever()
+
+    __asyncthread = threading.Thread(target=start, name="__telethon_async_thread__",
+                                     daemon=True)
+    __asyncthread.start()
+    __asyncthread.loop = loop
+    __asyncthread.executor = executor
+
+    TelegramClient.__init__ = functools.partialmethod(TelegramClient.__init__,
+                                                      loop=loop)
+
+    _syncify(TelegramClient, Draft, Dialog, MessageButton, ChatGetter,
+             SenderGetter, Forward, Message, InlineResult, Conversation,
+             loop=loop, thread_ident=__asyncthread.ident)
+    _syncify_wrap(TelegramClient, "start", loop, __asyncthread.ident)
+
+    old_add_event_handler = TelegramClient.add_event_handler
+    old_remove_event_handler = TelegramClient.remove_event_handler
+    proxied_event_handlers = {}
+
+    @functools.wraps(old_add_event_handler)
+    def add_proxied_event_handler(self, callback, *args, **kwargs):
+        def _proxy(*pargs, **pkwargs):
+            executor.submit(callback, *pargs, **pkwargs)
```

Comment:
This breaks the current propagation logic. Handlers for an event are supposed to be called sequentially and the `StopPropagation` should do exactly what it says. But this just adds the task to the executor and moves on discarding the result.